### ISSUE TITLE
lint: add modify-local-in-defer check

### ISSUE THREAD

**Issue by sergey-shambir** [go-critic/go-critic#801]
Our team faced with code that modifies variable in `defer` after `return` directive:
* [bad code](https://play.golang.org/p/FXJJOHzLoqc)
* [fixed code](https://play.golang.org/p/da3Ag_Zw1NL)

```go
// incorrect code - "one" never appended since defer modifies local variable
func test() ([]string) {
	result := make([]string, 0, 3)
	defer func() {
		result = append(result, "one")
	}()
	return result
}

// correct code - "one" appended since defer modifies named return value
func test() (result []string) {
	result = make([]string, 0, 3)
	defer func() {
		result = append(result, "one")
	}()
	return result
}
```

I'm not sure if there are any valid reasons to assign variables used with `return` inside `defer`.

**Comment by Quasilyte** [comment]
> I'm not sure if there are any valid reasons to assign variables used with `return` inside `defer`.

It's the only way to recover from panic and update values to be returned, AFAIK.
